### PR TITLE
Provide auto-init method

### PR DESCRIPTION
**Pull request by beigirad**
- Provide `API KEY` via manifest meta-data for reducing required arguments.
- Make the initialization easier for simple usages.
-- Use AppStartup library for auto-init trigger.
-- Provide another flag for advanced usages.
- Update readme

### CODE REVIEW

File: WeatherNotFound/build.gradle.kts [hunk 1]
```diff
@@ -48,6 +48,7 @@ dependencies {
     val appCompatVersion = "1.6.1"
 
     implementation("androidx.appcompat:appcompat:$appCompatVersion")
+    implementation("androidx.startup:startup-runtime:1.1.1")
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/build.gradle.kts:51]: As you can see, in the Gradle file, the style of dependencies is like the following:

1. define the library version via a variable. like `appCompatVersion`
2. then use it in `implementation` section.

I will appreciate change this line and implement `runtime` library using two steps above.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt [hunk 2]
```diff
@@ -78,7 +79,14 @@ class WeatherNotFound private constructor() {
...
-        OPEN_WEATHER_API_KEY = openWeatherApiKey
+        OPEN_WEATHER_API_KEY = context.packageManager
+            .getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
+            .metaData
+            .getString(OPEN_WEATHER_TOKEN_MANIFEST_KEY)
+            ?: throw IllegalArgumentException(
+                "$OPEN_WEATHER_TOKEN_MANIFEST_KEY doesn't found in manifest. " +
+                        "Please provide it as metadata in your application manifest"
+            )
         // Todo: For now, SDK can just support Json responses! Other formats will be added soon.
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:90]: Please create a function for this operation and return just a `String` as the result. I believe it will help the developer understand code more easily.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt [hunk 3]
```diff
@@ -66,7 +68,6 @@ class WeatherNotFound private constructor() {
      */
     fun init(
         context: Context,
-        openWeatherApiKey: String,
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:69]: Imagine developer doesn't want to initialize library in manifest at all, in this scenario, they expect to fully initialize library within java or kotlin code. at the end I think it's good to receive `openWeatherApiKey` from this function.

please revert the changes or comment your idea about this.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt [hunk 4]
```diff
@@ -105,6 +113,19 @@ class WeatherNotFound private constructor() {
         }
 
         validateOpenWeatherApiKey {}
+
+        Log.i(TAG, "initialized completely")
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:117]: Please remove the log. library shouldn't have logs in production.


> Maybe I write some logging functionality for debug mode later...

File: app/src/main/java/me/learning/app/MainActivity.kt [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package me.learning.app
 
 import android.os.Bundle
+import android.widget.Toast
```
Comment (AbolfaZlRezaEe) [inline app/src/main/java/me/learning/app/MainActivity.kt:4]: Please undo all changes of this file. Thanks.

**Review by AbolfaZlRezaEe — CHANGES_REQUESTED** [review]

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt [hunk 6]
```diff
@@ -78,7 +79,14 @@ class WeatherNotFound private constructor() {
 
         setInitCalledState(true)
 
-        OPEN_WEATHER_API_KEY = openWeatherApiKey
+        OPEN_WEATHER_API_KEY = context.packageManager
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:82]: Please create a function for this operation and return just an `String` as the result. I believe it will help the developer understand code more easily.

File: WeatherNotFound/build.gradle.kts (hunk 1 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/build.gradle.kts:51]: As you can see, in the Gradle file, the style of dependencies is like the following:
1. define the library version via a variable. like `appCompatVersion`
2. then use it in implementation section.


I will appreciate change this line and implement runtime library using two steps above.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt (hunk 3 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:69]: Imagine developer doesn't want to initialize library in manifest at all, in this scenario, they expect to fully initialize library within java or kotlin code. at the end I think it's good to receive openWeatherApiKey from this function.
please revert the changes or comment your idea about this.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt (hunk 4 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:117]: Please remove the log. library shouldn't have logs in production.

> Maybe I write some logging functionality for debug mode later..

File: app/src/main/java/me/learning/app/MainActivity.kt (hunk 5 above)
Comment (AbolfaZlRezaEe) [inline app/src/main/java/me/learning/app/MainActivity.kt:4]: Please undo all changes of this file. Thanks.

**Review by AbolfaZlRezaEe — CHANGES_REQUESTED** [review]

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt (hunk 3 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:69]: Imagine developer doesn't want to initialize library in manifest at all, in this scenario, they expect to fully initialize library within java or kotlin code. at the end I think it's good to receive openWeatherApiKey from this function.
please revert the changes or comment your idea about this.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt (hunk 6 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:82]: Please create a function for this operation and return just an `String`as the result. I believe it will help the developer understand code more easily.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt (hunk 4 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:117]: Please remove the log. library shouldn't have logs in production.

> Maybe I write some logging functionality for debug mode later..

File: WeatherNotFound/build.gradle.kts (hunk 1 above)
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/build.gradle.kts:51]: As you can see, in the Gradle file, the style of dependencies is like the following:

1. define the library version via a variable. like `appCompatVersion`
2. then use it in implementation section.


I will appreciate change this line and implement runtime library using two steps above.

File: app/src/main/java/me/learning/app/MainActivity.kt (hunk 5 above)
Comment (AbolfaZlRezaEe) [inline app/src/main/java/me/learning/app/MainActivity.kt:4]: Please undo all changes of this file. Thanks.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/initializer/WeatherNotFoundInitializer.kt [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package me.learning.weathernotfound.initializer
+
+import android.content.Context
+import androidx.startup.Initializer
+import me.learning.weathernotfound.presentation.WeatherNotFound
+
+
+class WeatherNotFoundInitializer : Initializer<WeatherNotFound> {
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/initializer/WeatherNotFoundInitializer.kt:8]: Because this class must reachable only inside of library, change the visibility of it to `internal`.

also put this class in `presentation` layer.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt [hunk 8]
```diff
@@ -66,7 +68,6 @@ class WeatherNotFound private constructor() {
      */
     fun init(
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:69]: We have a problem here: Let's assume I wrote the `auto-init` metadata and set it `true`, after that I called `init` function also in `BaseApplication`. in this scenario, we init library twice!

For this problem, we should throw an exception for developers to understand that they initialized library twice and it is wrong!

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt [hunk 9]
```diff
@@ -105,6 +113,19 @@ class WeatherNotFound private constructor() {
...
         validateOpenWeatherApiKey {}
+
+        Log.i(TAG, "initialized completely")
+    }
+
+    fun tryAutoInit(context: Context) {
+        val canAutoInit = context.packageManager
+            .getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
+            .metaData
+            .getBoolean(AUT_INIT_MANIFEST_KEY, true)
```
Comment (AbolfaZlRezaEe) [inline WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:124]: I believe until developers didn't write `auto-init` and set it to `true` or `false`, we **Shouldn't** initialize library.

So with this definition, `defaultValue` of this function should be `false`.

File: WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt (hunk 3 above)
Comment (beigirad) [reply WeatherNotFound/src/main/java/me/learning/weathernotfound/presentation/WeatherNotFound.kt:69]: This PR is based on this supposition that reducing setup steps and providing this key as a manifest meta-data is first commit and I can't revert it.

So I close this PR
